### PR TITLE
Adsk Contrib - Add bit-depth support to apply methods

### PR DESCRIPTION
**Pull request by hodoulp**
The pull request implements an important part of #609 .

> OCIO currently only accepts floating-point image buffers. However, in many scenarios one is starting with an integer buffer. Furthermore, the first operation often performed is a 1d-LUT. So rather than accepting an integer buffer and doing a fast look-up, OCIO currently converts to float and does interpolation rather than a simple look-up.

Some ground works were already done to ease the bit-depth support integration and reduce the final pull request size. Please refer to #699 and #722 .

> When a 1d-LUT is the first operator in the chain, it will be interpolated to the needed bit-depth and then a look-up will be done on the pixels. This gives a big performance boost with no loss of precision.

The implemented approach is now using the first (or last) op in the color transform (i.e. processor instance) to perform the input and output bit-depth conversions when needed.

**Public API:**
To remove the existing limitation of the OCIOv1 API, the new design provides several improvements:
- Ability to specify an input and output ImageDesc supporting bit-depths:
`void CPUProcessor::apply(ImageDesc & imgDesc) const;`
`void CPUProcessor::apply(const ImageDesc & srcImgDesc, ImageDesc & dstImgDesc) const;`
- Ability to control the optimization (what optimization(s) to use or not), and finalization (fast or exact inverse LUT support):
`ConstCPUProcessorRcPtr Processor::getDefaultCPUProcessor() const;`
`ConstCPUProcessorRcPtr Processor::getOptimizedCPUProcessor(OptimizationFlags oFlags, FinalizationFlags fFlags) 
 const;`
- Ability to process diffferent input and output bit-depths:
`ConstCPUProcessorRcPtr Processor::getOptimizedCPUProcessor(BitDepth inBitDepth, BitDepth outBitDepth, OptimizationFlags oFlags,  FinalizationFlags fFlags) const;`

**Reviewers:**
I would suggest you start from the public API before digging in the implementation. One could start with the OpenColorIO.h file, the Config class (header + implementation), the Processor class (header + implementation), and the new CPUProcessor class (header + implementation and associated unit tests).

**Pseudo-code** (from ocioconvert):
`OIIO::ImageSpec spec;`
`OCIO::ImgBuffer img;`
`...`
`img.allocate(spec);`
`...`
`OCIO::ConstConfigRcPtr config = OCIO::GetCurrentConfig();`
`OCIO::ConstProcessorRcPtr processor = config->getProcessor(...);`
`const OCIO::BitDepth bitDepth = OCIO::GetBitDepth(spec);`
`OCIO::ConstCPUProcessorRcPtr cpuProcessor = processor->getOptimizedCPUProcessor(...);`
`OCIO::ImageDescRcPtr imgDesc = OCIO::CreateImageDesc(spec, img);`
`cpuProcessor->apply(*imgDesc);`

**Measures (RB-1.1 vs branch)** using spi-vfx:

| Conversion  | bit-depths | Improvement |
| ------------- | ------------- | ------------- |
| lnf to lgf  | F32 to F32  | **3x** |
| lgf to lnf  | F32 to F32  | **4x** |
| lnf to lgf  | UI16 to UI16 | **9.7x** |
| lgf to lnf  | UI16 to UI16 | **7x** |
| lnf to xyz16 | UI16 to UI16 | **3.4x** |

### CODE REVIEW

**Comment by hodoulp** [comment]
The finalization flag provides a way to select the `fast` or `exact` inversion mode (refer to #621).

**Comment by hodoulp** [comment]
The pull request fixes #401 .

**Comment by hodoulp** [comment]
In order to help everyone investigating and/or tracking performance, the pull request adds:
1. A new app named `ocioperf` to perform some predefined benchmarks i.e. full image, line-by-line and pixel-per-pixel,
2. A CPU processing measure in verbose mode to `ocioconvert`.

**Comment by scoopxyz** [comment]
This is terribly old, sorry Patrick.

What is currently blocking this? Just reviews?

**Comment by hodoulp** [comment]
@scoopxyz  Thanks, only waiting reviews. You could start from the public API to have an overview.

File: src/OpenColorIO/GPUProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,299 @@
+/*
```
Comment (michdolan) [inline src/OpenColorIO/GPUProcessor.cpp]: General question. Some file/class names use the "GPU" casing, and others use "Gpu". Should we standardise on one?

Comment (hodoulp) [reply src/OpenColorIO/GPUProcessor.cpp]: Good catch, the acronyms writing is quite inconsistent. Yes, we will have to update the coding style, and later have clang-tidy running in the CI build.

**Comment by hodoulp** [comment]
Following the two-weeks rule, here is a friendly reminder that the branch could now be merged this week unless there are some objections.

**Pull request merged**